// react-chart-editor/lib/components/containers/TraceRequiredPanel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _PanelEmpty = require('./PanelEmpty');

var _PanelEmpty2 = _interopRequireDefault(_PanelEmpty);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _derived = require('./derived');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var TraceRequiredPanel = function (_Component) {
  _inherits(TraceRequiredPanel, _Component);

  function TraceRequiredPanel() {
    _classCallCheck(this, TraceRequiredPanel);

    return _possibleConstructorReturn(this, (TraceRequiredPanel.__proto__ || Object.getPrototypeOf(TraceRequiredPanel)).apply(this, arguments));
  }

  _createClass(TraceRequiredPanel, [{
    key: 'hasTrace',
    value: function hasTrace() {
      return this.context.fullData.filter(function (trace) {
        return trace.visible;
      }).length > 0;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _ = this.context.localize;

      var _props = this.props,
          children = _props.children,
          rest = _objectWithoutProperties(_props, ['children']);

      var showPanel = true;
      var emptyPanelMessage = { heading: '', message: '' };

      var noTraceMessage = {
        heading: _("Looks like there aren't any traces defined yet."),
        message: _("Go to the 'Create' tab to define traces.")
      };

      var conditions = [function () {
        return _this2.hasTrace();
      }].concat(this.props.extraConditions ? this.props.extraConditions : []);

      var messages = [noTraceMessage].concat(this.props.extraEmptyPanelMessages ? this.props.extraEmptyPanelMessages : []);

      if (this.props.visible) {
        conditions.forEach(function (condition, index) {
          if (!showPanel) {
            return;
          }
          if (!condition()) {
            showPanel = false;
            emptyPanelMessage.heading = messages[index].heading;
            emptyPanelMessage.message = messages[index].message;
          }
        });

        if (showPanel) {
          return _react2.default.createElement(
            _derived.LayoutPanel,
            rest,
            children
          );
        }

        return _react2.default.createElement(_PanelEmpty2.default, {
          heading: emptyPanelMessage.heading,
          message: emptyPanelMessage.message
        });
      }
      return null;
    }
  }]);

  return TraceRequiredPanel;
}(_react.Component);

TraceRequiredPanel.propTypes = {
  children: _propTypes2.default.node,
  visible: _propTypes2.default.bool,
  extraConditions: _propTypes2.default.array,
  extraEmptyPanelMessages: _propTypes2.default.array
};

TraceRequiredPanel.defaultProps = {
  visible: true
};

TraceRequiredPanel.contextTypes = {
  fullData: _propTypes2.default.array,
  localize: _propTypes2.default.func
};

exports.default = TraceRequiredPanel;
